### DISCUSSION TITLE
Migration executing in "public" tenant despite not being in APPS

### DISCUSSION THREAD

**Discussion by vikingvynotking** [lorinkoz/django-pgschemas#118] [General]
Hi there!

I am in the process of lifting a py3.7/ django 3.2/ pgschemas 0.4.3 project up to py3.10/ django 4.1/ pgschemas 0.10.0 .

In the process, django wanted to create a bunch of new migrations for each app to update ID fields. No problem, I thought, except I'm now receiving the following error:

```
bash-5.1# ./manage.py migrate -s public
[sequential:public] Operations to perform:
[sequential:public]   Apply all migrations: accounts, admin, auth, authtoken, foo, contenttypes, django_celery_beat, bar, baz, boz, sessions, frob
[sequential:public] Running migrations:
[sequential:public]   Applying accounts.0003_auto_20220822_1143...Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedTable: relation "authtoken_token" does not exist
``` 

This is odd because the app that defines that model is not included in the `public` schema:

```
TENANTS = {
        'public': {
                'APPS': [
                        'django.contrib.contenttypes',
                        'django.contrib.staticfiles',
                        'django.contrib.gis',
                        'django.contrib.auth',
                        'django.contrib.admin',
                        'django.contrib.sessions',

                       ...,

                        'rest_framework',
                        'rest_framework_gis',
#                        'rest_framework.authtoken',   see I am commented out

                       ...
                ],

                'TENANT_MODEL': 'patrons.Patron',
                'DOMAIN_MODEL': 'patrons.Domain',
        },

```

If I run `manage.py migrate -s <my-tenant>` everything works ok. Another odd part of all this is that authtoken_token isn't even mentioned in any `accounts` migrations, but evidently it defined its own ID-update migration which is where the error originates. So I'm not sure why it is getting reported by *my* migration.

Any ideas what I might be doing wrong? I am considering creating an interim release which includes rest_framework.authtoken as a separate migration step, but before going down that path I'm wondering if there's a better, cleaner approach.

Thanks in advance!

**Comment by lorinkoz** [comment]
Is it because the migration depends on the missing package?

**Comment by vikingvynotking** [comment]
Thanks for the quick response!

It shouldn't/ doesn't appear to. `accounts` just stores user names and such, and has nothing to do with how those users authenticate.

BTW I just ran this:

```
./manage.py migrate authtoken zero --fake -s public
./manage.py migrate -s public
```

and everything passed, including running the authtoken migrations. It still makes no sense since authtoken isn't installed for `public`. Am I misunderstanding something?


```
bash-5.1# ./manage.py migrate authtoken zero --fake -s public
[sequential:public] Operations to perform:
[sequential:public]   Unapply all migrations: authtoken
[sequential:public] Running migrations:
[sequential:public]   Rendering model states... DONE
[sequential:public]   Unapplying authtoken.0003_tokenproxy... FAKED
[sequential:public]   Unapplying authtoken.0002_auto_20160226_1747... FAKED
[sequential:public]   Unapplying authtoken.0001_initial... FAKED
bash-5.1# ./manage.py migrate -s public
[sequential:public] Operations to perform:
[sequential:public]   Apply all migrations: accounts, admin, auth, authtoken, ...
[sequential:public] Running migrations:
[sequential:public]   Applying accounts.0003_auto_20220822_1143... OK
[sequential:public]   Applying auth.0012_alter_user_first_name_max_length... OK
[sequential:public]   Applying authtoken.0001_initial... OK
[sequential:public]   Applying authtoken.0002_auto_20160226_1747... OK
[sequential:public]   Applying authtoken.0003_tokenproxy... OK
```

and to further confuse me, if accounts depended on authtoken I would expect it to run after, not before, those migrations.

**Comment by lorinkoz** [comment]
If accounts depends on authtoken, migrations for authtoken will run first.

**Comment by vikingvynotking** [comment]
Maybe I'm not being clear: accounts does not depend on authtoken, anywhere.

```
$ ack authtoken accounts/
# nothing returned
```

but even if it did, `authtoken` migrations would run before `accounts`, and they don't - see my previous message.

But really the problem is since authtoken isn't installed for public, I wouldn't expect its migrations to run there and fail.

**Reply by lorinkoz** [reply]
Are you using a custom user model? If yes, what's the app for it?

**Comment by vikingvynotking** [comment]
`authtoken` migrations show as having run in public, but the tables are not there (which is correct). But if they are run and don't create tables, why would a later alter table operation fail? Seems like it should follow the same path, right? Or does pgschemas somehow delete unused tables? I confess I've not dug too deeply into the db management code inside this library.

Edit: to clarify, the migrations show as having run in an older instance of this project, i.e. before the upgrade.
